### ISSUE TITLE
Empty rows should not be rendered

### ISSUE THREAD

**Issue by ericscheid** [naturalcrit/homebrewery#1729]
Consider this table, where very long text data has been split over multiple lines for editing reasons, and all the cells of the row are rowspanned into the above row..

````
| Race     | Multi-class options                                         |
|:---------|:------------------------------------------------------------|
| Dwarf    | fighter/cleric, fighter/thief                               |
| Elf      | fighter/mage, fighter/thief, mage/thief,                    |
|         ^| fighter/mage/thief                                         ^|
| Gnome    | fighter/thief, fighter/cleric, fighter/illusionist,         | 
|         ^| thief/cleric, thief/illusionist, cleric/illusionist        ^|
| Half-elf | fighter/priest, fighter/mage, fighter/thief, ranger/priest, |
|         ^| mage/priest, mage/thief,                                   ^|
|         ^| fighter/mage/priest, fighter/mage/thief                    ^|
| Halfling | fighter/thief                                               |
| Human    | none                                                        |
````

Unfortunately, the markdown renders this with empty `<tr></tr>`...

![image](https://user-images.githubusercontent.com/2605467/135560819-45ef0cdd-0021-4858-9562-1ecaf64cf634.png)

This is a bug, and it messes up the even/odd row striping:

![image](https://user-images.githubusercontent.com/2605467/135560873-f9d01e19-7849-425c-b7fe-e64c516594db.png)

**Comment by calculuschild** [comment]
Hmm interesting find. Good eye.

**Comment by G-Ambatte** [comment]
To further muddy the waters, if the `<tr></tr>` is removed completely, then the preceding `rowspan`(s) must also be decremented, or else the entire table format will break in new and interesting ways.

As I understand it, this is not actually the intended use of `rowspan`. I can see how it might be clearer in the editor and thus more convenient - but as per the current implementation, it doesn't actually work.

**WORKAROUNDS:**
* If the user doesn't attempt to misuse `rowspan` in this manner and keeps the content for a single row in a single row, then it works fine.
![image](https://user-images.githubusercontent.com/19826659/135741897-5a42d1d1-da3e-4cb1-9996-72936ee6a6ed.png)
* If the user ensures that they always span an odd number of rows (for the entire row), then it works fine.
![image](https://user-images.githubusercontent.com/19826659/135741825-d883e058-7e98-4cda-bd98-8bcbb29deaba.png)

It's worth noting that, despite the different approaches, the two tables above appear to be visually identical.

**Comment by ericscheid** [comment]
Oh, good point on decrementing rowspans.

**Comment by dbolack-ab** [comment]
Would the simplest fix be to strip empty <tr></tr> pairs in html.renderer()?

Or must this be fixed in marked-extended-tables ?

**Comment by calculuschild** [comment]
Fixes to table generation should go in marked-extended-tables

**Comment by dbolack-ab** [comment]
In addition to the two workarounds above, there is one *breaking* fix.

This is to wrap rows into sections using multiple `<tbody>` and adjust the CSS to nth-child on the `<tbody>` instead of the `<tr>`

**Comment by calculuschild** [comment]
> In addition to the two workarounds above, there is one *breaking* fix.
> 
> This is to wrap rows into sections using multiple `<tbody>` and adjust the CSS to nth-child on the `<tbody>` instead of the `<tr>` 
> 
> 

Agreed, this would be breaking for most of our users. Let's not do this.

**Comment by dbolack-ab** [comment]
Demonstration of "fix" 

https://github.com/calculuschild/marked-extended-tables/pull/466 with https://github.com/naturalcrit/homebrewery/pull/4002

**Comment by dbolack-ab** [comment]
I have updated the fix in what I think is an adequately backwards compatible way.

The marked module adds a toggle that switches the render between the current behavior and the "fix" behavior.

To use the toggle, I added the option of passing the renderer version through to Markdown.render() ( assumes V3 ) and using that to set the toggle. This is visible in #4002.

This fix *does* assume a V4 rendering engine exists and the CSS is updated....

**Comment by calculuschild** [comment]
I see what this fix is going for, but looking over the source I think it shouldn't be too hard to make a fix that just removes empty rows from the final output. Then we don't have to worry about backwards compatibility with extra `<tbody>` at all.

Because one thing to consider is that extended-tables package is used by other projects outside of the Homebrewery. Meaning, we should aim to fix things in a way that everyone can get the fix without having to enter a secret toggle, and that it remains in spec with GFM as much as possible (single tbody, etc.). Not to mention, adding a toggle means we now need to maintain and test *two* versions of the extension.

I think it might be as simple as, before emitting a `<tr>` in the renderer part, check if all the cells have rowspan 0. If yes, don't render the row at all.

**Comment by dbolack-ab** [comment]
> I see what this fix is going for, but looking over the source I think it shouldn't be too hard to make a fix that just removes empty rows from the final output. Then we don't have to worry about backwards compatibility with extra `<tbody>` at all.
> 
> Because one thing to consider is that extended-tables package is used by other projects outside of the Homebrewery. Meaning, we should aim to fix things in a way that everyone can get the fix without having to enter a secret toggle, and that it remains in spec with GFM as much as possible (single tbody, etc.). Not to mention, adding a toggle means we now need to maintain and test _two_ versions of the extension.
> 
> I think it might be as simple as, before emitting a `<tr>` in the renderer part, check if all the cells have rowspan 0. If yes, don't render the row at all. Possibly also check the next row if it is empty and adjust the current row span as needed.

No, that won't work.

If you Have rowspans, but do NOT include `<tr>` elements for them to traverse into Chrome gets *really*  confused and renders the table wrong.

```
<table><thead><tr><th align="left">Race</th>
<th align="left">Multi-class options</th>
</tr></thead><tbody><tr><td align="left">Dwarf</td>
<td align="left">fighter/cleric, fighter/thief</td>
</tr><tr><td rowspan="2" align="left">Elf </td>
<td rowspan="2" align="left">fighter/mage, fighter/thief, mage/thief, fighter/mage/thief                                         </td>
</tr><tr><td rowspan="2" align="left">Gnome </td>
<td rowspan="2" align="left">fighter/thief, fighter/cleric, fighter/illusionist, thief/cleric, thief/illusionist, cleric/illusionist        </td>
</tr><tr><td rowspan="3" align="left">Half-elf  </td>
<td rowspan="3" align="left">fighter/priest, fighter/mage, fighter/thief, ranger/priest, mage/priest, mage/thief,                                    fighter/mage/priest, fighter/mage/thief                    </td>
</tr>

<tr><td align="left">Halfling</td>
<td align="left">fighter/thief</td>
</tr><tr><td align="left">Human</td>
<td align="left">none</td>
</tr></tbody></table>
```

![Image](https://github.com/user-attachments/assets/e5ca8e61-cfa7-448a-b389-7cf5bf3e2f2a)

**Comment by G-Ambatte** [comment]
Might be best as a walkTokens function, to check for rowspans that would cross into empty rows, then decrement that rowspan by 1 before removing the empty row?

**Comment by calculuschild** [comment]
> No, that won't work.
> 
> If you Have rowspans, but do NOT include <tr> elements for them to traverse into Chrome gets really confused and renders the table wrong.

I suspect you did not read my last paragraph fully.

**Comment by G-Ambatte** [comment]
> This should be doable directly in the Renderer part.

I had made some assumptions which I realize now were wrong. I suspect that attempting to do this in `walkTokens` would be a very large task.

**Comment by G-Ambatte** [comment]
As I understand the function, `splitCells` walks the table row string and breaks it into individual cells; if the cell is to be merged, the object has a `rowSpanTarget` property. So if the entire row has no data in the output string (`<tr></tr>`), then the entire array of `cells` can be walked and all `rowSpanTarget` properties decremented before removing the string.

![Image](https://github.com/user-attachments/assets/130f9e64-1fe5-41c3-86aa-6874414c121e)

Actually, that solution would be in the Tokenizer, rather than the Renderer.

**Comment by G-Ambatte** [comment]
> I think it might be as simple as, before emitting a <tr> in the renderer part, check if all the cells have rowspan 0. If yes, don't render the row at all. Possibly also check the next row if it is empty and adjust the current row span as needed.

From the small amount of poking at it this afternoon, I believe that this is the best way.

**Comment by G-Ambatte** [comment]
It wouldn't take much effort to add a configuration option - e.g. `skipEmptyRows : false | true` - so the functionality is under the control of the user.

**Comment by G-Ambatte** [comment]
In the latest commit, I've added a configuration object with the `skipEmptyRows` property to https://github.com/calculuschild/marked-extended-tables/pull/474, which defaults to `false`, preserving the current operation. At this stage, I have not rolled the `endRegex` parameter into the configuration object, as this would change any current implementation which uses that parameter.
